src/Pages/ProductDetails.js

```diff
@@ -0,0 +1,153 @@
+import React from "react";
+import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";
+import { Link } from "react-router-dom";
+import { useNavigate } from "react-router";
+import Loader from "react-loader-spinner";
+import { useParams } from "react-router";
+import { useCart } from "../Context/cart-context";
+import { useAuth } from "../Context/AuthProvider";
+import { Toast } from "../components/Toast";
+import { addToServer, removeFromServer } from "../api/ServerHandler";
+export const ProductDetails = () => {
+  const { productId } = useParams();
+  const navigate = useNavigate();
+  const { isUserLogin } = useAuth();
+  const { showToast, itemsInCart, dispatch } = useCart();
+  function getProduct(products, productId) {
+    return products.find((product) => product._id === productId);
+  }
+  async function cartHandler() {
+    if (isUserLogin) {
+      dispatch({ type: "SHOW_TOAST", payload: "Adding to Cart" });
+      await addToServer("cart", selectedProduct);
+      dispatch({ type: "ADD_TO_CART", payload: selectedProduct });
```

Comment:
What if the API call fails? It will still the product to the state, right? I think it will be better if these success dispatch are kept inside the if - else block for sucess part, else redirect the user to error page or show the error in UI.